[lldb] Apply available search paths from swift to clang importer

cachemeifyoucan opened the pull request:
For explicit build that uses direct cc1 options (like swift caching), all
search paths are dropped after dependency scanning. The clang modules
from the explicit build are the only modules available for lldb, which
should not be a problem if the context for expression evaluator in lldb
matches the context for swift compiler. But there are currently, there
are edge cases, like when lldb is constructing a type from generics from
a different module, it might tries to import additional clang modules.

Teach lldb to apply the swift search paths to clang importer, in case
some implicit clang module build is required, to make the debugging
process less disruptive. This cannot recover all the search paths (like
clang specific ones that passes through -Xcc options), but lldb were
already relying on the search paths from current module context can find
the clang modules from a different context. This should make the
behavior closer to implicit build before the lldb context and swift
compiler context can match perfectly.

rdar://138664252

cachemeifyoucan committed:
[lldb] Allow swift-clang-override-options setting when using cc1 args

When using swift caching or direct cc1 mode, allow using setting to
override clang arguments. Even this override option directly applies to
cc1 arguments, not clang driver arguments, thus is a behavior change,
but it can be handy for debugging or working around issues.

cachemeifyoucan committed:
[lldb] Apply available search paths from swift to clang importer

For explicit build that uses direct cc1 options (like swift caching), all
search paths are dropped after dependency scanning. The clang modules
from the explicit build are the only modules available for lldb, which
should not be a problem if the context for expression evaluator in lldb
matches the context for swift compiler. But there are currently, there
are edge cases, like when lldb is constructing a type from generics from
a different module, it might tries to import additional clang modules.

Teach lldb to apply the swift search paths to clang importer, in case
some implicit clang module build is required, to make the debugging
process less disruptive. This cannot recover all the search paths (like
clang specific ones that passes through -Xcc options), but lldb were
already relying on the search paths from current module context can find
the clang modules from a different context. This should make the
behavior closer to implicit build before the lldb context and swift
compiler context can match perfectly.

rdar://138664252

cachemeifyoucan commented:
@swift-ci please test

cachemeifyoucan commented:
@swift-ci please test

cachemeifyoucan commented:
@swift-ci please test windows platform

cachemeifyoucan commented:
@swift-ci please test

cachemeifyoucan commented:
@swift-ci please test

cachemeifyoucan commented:
Patch updated since there are more work needed to get implicit module build, mostly how to configure module cache path early enough.

Also it is tricky to get some default search path (since we don't run clang driver). Might have some room for improvement there.

cachemeifyoucan commented:
Please test with following pull request:
https://github.com/swiftlang/swift/pull/77485

@swift-ci please test

adrian-prantl commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.h:
```diff
@@ -946,6 +958,8 @@ class SwiftASTContext : public TypeSystemSwift {
   /// A cache for GetCompileUnitImports();
   llvm::DenseSet<std::pair<Module *, lldb::user_id_t>> m_cu_imports;
+  /// clang module cache path.
+  std::string m_module_cache_path = "";
```
Why do we need to store this separately? Isn't it sufficient to compute it once and then only have it stored in the CompilerInvocation?

cachemeifyoucan replied:
This is prevents the path gets computed every time a context is requested. Is there a better place for it or you prefer to re-compute every time?

adrian-prantl replied:
Each SwiftASTContext owns exactly one CompilerInvocation, there is nothing gained by storing m_module_cache_path in SwiftASTContext.

adrian-prantl replied:
But you can sink the code that generates the tmpdir into GetClangModulesCacheProperty, and potentially avoid running that if GetClangModulesCachePath didn't change (by storing the last result in a static variable).

cachemeifyoucan replied:
I think I figured out. Let me know what you think now.

cachemeifyoucan committed:
[lldb] Improve swift clang importer when explicit module missing

Instead of dropping all explicit modules when found one module is
missing, only drop the ones that are missing. In this case, the implicit
build will only fill in the missing ones, rather than rebuild
everything.

cachemeifyoucan commented:
@swift-ci please test

cachemeifyoucan commented:
Please test with following pull request:
https://github.com/swiftlang/swift/pull/77485

@swift-ci please test
